src/scatter.jl

```diff
@@ -0,0 +1,187 @@
+                  src::AbstractArray{Tsrc,Nsrc},
+                  idx::AbstractArray{<:Integer,Nidx}) where {Tdst,Tsrc,Ndst,Nsrc,Nidx}
+    Ns = scatter!(+, zero(dst), one.(src), idx)
+    dst_ = scatter!(+, zero(dst), src, idx)
+    dst .+= safe_div.(dst_, Ns)
+    return dst
+end
+
+function scatter!(op::typeof(mean),
+                  dst::AbstractArray{Tdst,Ndst},
+                  src::AbstractArray{Tsrc,Nsrc},
+                  idx::AbstractArray{NTuple{N,Int},Nidx}) where {Tdst,Tsrc,Ndst,Nsrc,N,Nidx}
+    Ns = scatter!(+, zero(dst), one.(src), idx)
+    dst_ = scatter!(+, zero(dst), src, idx)
+    dst .+= safe_div.(dst_, Ns)
+    return dst
+end
+
+
+"""
+    scatter(op, src, idx)
+
+Scatter operation, which applies specified operation on `src` according to `idx`
+and gives an new array `dst`.
+For each index `k` in `idx`, accumulate values in `dst` according to
+
+    dst[idx[k]...] = (op).(src[k...])
+
+# Arguments
+- `op`: operations to be applied on `dst` and `src`, e.g. `+`, `-`, `*`, `/`, `max` and `min`.
+- `src`: the source data for aggregating.
+- `idx`: the mapping for aggregation from source (index) to destination (value).
+The index of `idx` is corresponding to the index of `src` and the value of `idx` is
+corresponding to the index of `dst`. The value of `idx` can be `Int` or `Tuple` type.
+"""
+function scatter end
+
+for op in [+, -]
+    @eval function scatter(op::typeof($op),
+                           src::AbstractArray{T,Nsrc},
+                           idx::AbstractArray{<:IntOrTuple,Nidx}) where {T,Nsrc,Nidx}
+        dims = Nsrc - Nidx
+        dst = zeros(T, size(src)[1:dims]..., maximum_dims(idx)...)
+        scatter!(op, dst, src, idx)
+    end
+end
+
+for op in [*, /]
+    @eval function scatter(op::typeof($op),
+                           src::AbstractArray{T,Nsrc},
+                           idx::AbstractArray{<:IntOrTuple,Nidx}) where {T,Nsrc,Nidx}
+        dims = Nsrc - Nidx
+        dst = ones(T, size(src)[1:dims]..., maximum_dims(idx)...)
+        scatter!(op, dst, src, idx)
+    end
+end
+
+function scatter(op::typeof(max),
+                 src::AbstractArray{T,Nsrc},
+                 idx::AbstractArray{<:IntOrTuple,Nidx}) where {T,Nsrc,Nidx}
+    dims = Nsrc - Nidx
+    dst = fill(typemin(T), size(src)[1:dims]..., maximum_dims(idx)...)
+    scatter!(op, dst, src, idx)
+end
+
+function scatter(op::typeof(min),
+                 src::AbstractArray{T,Nsrc},
+                 idx::AbstractArray{<:IntOrTuple,Nidx}) where {T,Nsrc,Nidx}
+    dims = Nsrc - Nidx
+    dst = fill(typemax(T), size(src)[1:dims]..., maximum_dims(idx)...)
+    scatter!(op, dst, src, idx)
+end
+
+function scatter(op::typeof(mean),
+                 src::AbstractArray{T,Nsrc},
+                 idx::AbstractArray{<:IntOrTuple,Nidx}) where {T,Nsrc,Nidx}
+    FT = float(T)
+    dims = Nsrc - Nidx
+    dst = zeros(FT, size(src)[1:dims]..., maximum_dims(idx)...)
+    scatter!(op, dst, FT.(src), idx)
+end
```

Comment:
Instead of all these methods, one can use just this single one (or somthing similar)
```julia
function scatter(op, 
                src::AbstractArray{Tsrc, Nsrc}, 
                idx::AbstractArray{Tidx, Nidx}) where 
                    {Tsrc, Nsrc, Nidx, Tidx<:IntOrIntTuple}

    M = typelength(Tidx) 
    dstsize = (size(src)[1:Nsrc-M]..., size(idx)...)
    dst = similar(src, Tsrc, dstsize)
    fill!(dst, Base.reduce_empty(op, Tsrc))
    return scatter!(op, dst, src, idx)
end

typelength(::Type{<:Number}) = 1
typelength(::Type{<:NTuple{M}}) where M = M
```

Reply:
actually not a good idea, `Base.reduce_empty` doesn't support min/max. 
But please use `similar` and `fill!` instead of `zeros` and `ones` so that we support generic arrays.

Reply:
Oh! Doesn't `zeros` and `ones` support generic arrays?

Reply:
It does afaik

It needs the identities for the eltypes defined.

Reply:
no they don't, they will always return `Array` type, to support generic arrays (e.g. StaticArray, CuArray,...) the correct thing to do is `similar` and then `fill!`

Reply:
```julia
julia> struct M{T,S}
         a::T
         b::S
       end

julia> mr = reshape([M(rand(), rand()) for _ = 1:9], 3,3)
3×3 Array{M{Float64,Float64},2}:
 M{Float64,Float64}(0.33762, 0.0495362)  …  M{Float64,Float64}(0.260097, 0.922337)
 M{Float64,Float64}(0.282951, 0.18587)      M{Float64,Float64}(0.487388, 0.81491)
 M{Float64,Float64}(0.808664, 0.79451)      M{Float64,Float64}(0.0642862, 0.964739)

julia> Base.zero(::Type{<:M{T,S}}) where {T,S} = M(zero(T), zero(S))

julia> zero(mr)
3×3 Array{M{Float64,Float64},2}:
 M{Float64,Float64}(0.0, 0.0)  …  M{Float64,Float64}(0.0, 0.0)
 M{Float64,Float64}(0.0, 0.0)     M{Float64,Float64}(0.0, 0.0)
 M{Float64,Float64}(0.0, 0.0)     M{Float64,Float64}(0.0, 0.0)

julia> @code_warntype zero(mr)
Variables
  #self#::Core.Compiler.Const(zero, false)
  x::Array{M{Float64,Float64},2}

Body::Array{M{Float64,Float64},2}
1 ─ %1 = Base.similar(x)::Array{M{Float64,Float64},2}
│   %2 = Base.zero($(Expr(:static_parameter, 1)))::Core.Compiler.Const(M{Float64,Float64}(0.0, 0.0), false)
│   %3 = Base.fill!(%1, %2)::Array{M{Float64,Float64},2}
└──      return %3

```

Seems like it works as expected

Reply:
For `maximum_dims`, I want the dimensions of the value in `idx`, not the dimensions of `idx` itself. So, `size(idx)` may not work for my case.

Reply:
With the automatic lowering for the different types of arrays

Reply:
Done.